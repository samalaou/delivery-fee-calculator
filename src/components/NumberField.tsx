import { Grid, InputAdornment } from "@mui/material";
import TextField from "@mui/material/TextField";
import { ChangeEvent } from "react";
import { NumericFormat } from "react-number-format";

interface NumberFieldProps {
  dataTestId: string;
  error: boolean;
  decimalScale?: number;
  endAdornment?: string;
  id: string;
  label: string;
  name: string;
  value: number | "";
  handleChange(e: ChangeEvent<HTMLInputElement>): void;
}

function NumberField(props: NumberFieldProps): JSX.Element {
  // Used react-number-format with TextField to solve type="number" issues
  // https://mui.com/material-ui/react-text-field/#type-quot-number-quot
  return (
    <Grid item xs={12} style={{ margin: "20px 10px" }}>
      <NumericFormat
        id={props.id}
        customInput={TextField}
        label={props.label}
        value={props.value}
        allowLeadingZeros={false}
        decimalSeparator=","
        margin="none"
        decimalScale={props.decimalScale ? props.decimalScale : undefined}
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              {" "}
              {props.endAdornment || ""}
            </InputAdornment>
          ),
          inputProps: {
            min: 0,
            "data-test-id": props.dataTestId,
            name: props.name,
            notched: "true",
          },
        }}
        onChange={(e) => props.handleChange(e)}
        fullWidth
        required
        error={props.error}
        helperText={props.error && "This field must contain a valid number"}
        FormHelperTextProps={{
          sx: {
            backgroundColor: "primary.main",
            padding: 0,
            margin: 0,
          },
        }}
      />
    </Grid>
  );
}

export default NumberField;
