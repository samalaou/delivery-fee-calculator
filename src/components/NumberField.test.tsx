import { render, fireEvent } from "@testing-library/react";
import NumberField from "./NumberField";

describe("NumberField Component", () => {
  const handleChangeMock = jest.fn();

  beforeEach(() => {
    handleChangeMock.mockClear();
  });

  it("renders correctly", () => {
    const { getByTestId } = render(
      <NumberField
        id="test"
        value={10}
        handleChange={handleChangeMock}
        dataTestId="test-input"
        name="test"
        label="Test Label"
        error={false}
      />,
    );
    expect(getByTestId("test-input")).toBeInTheDocument();
  });

  it("matchs snapshot", () => {
    const { container } = render(
      <NumberField
        id="test"
        value={10}
        handleChange={handleChangeMock}
        dataTestId="test-input"
        name="test"
        label="Test Label"
        error={false}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("calls handleChange when input value changes", () => {
    const { getByTestId } = render(
      <NumberField
        id="test-id"
        label="Test Label"
        name="test-name"
        value={10}
        handleChange={handleChangeMock}
        dataTestId="test-data-testid"
        error={false}
      />,
    );

    fireEvent.change(getByTestId("test-data-testid"), {
      target: { value: "20" },
    });

    expect(handleChangeMock).toHaveBeenCalledTimes(1);
    expect(handleChangeMock).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ value: "20" }),
      }),
    );
  });

  it("applies correct styling to FormHelperText when error is true", () => {
    const { getByText } = render(
      <NumberField
        id="test"
        value={10}
        handleChange={handleChangeMock}
        dataTestId="test-input"
        name="test"
        label="Test Label"
        error={true} // Set error to true to display helper text
      />,
    );
    const formHelperText = getByText("This field must contain a valid number");
    expect(formHelperText).toHaveStyle("padding: 0");
    expect(formHelperText).toHaveStyle("margin: 0");
  });
});
