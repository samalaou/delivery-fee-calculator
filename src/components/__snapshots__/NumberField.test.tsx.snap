// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberField Component matchs snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
    style="margin: 20px 10px;"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      inputmode="numeric"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="test"
        id="test-label"
      >
        Test Label
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          data-test-id="test-input"
          id="test"
          min="0"
          name="test"
          notched="true"
          required=""
          type="text"
          value="10"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
           
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Test Label
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
